engine/src/main/java/org/apache/apex/log/appender/ApexRollingFileAppender.java

```diff
@@ -0,0 +1,174 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.log.appender;
+
+import java.io.BufferedWriter;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+import org.apache.log4j.RollingFileAppender;
+import org.apache.log4j.helpers.CountingQuietWriter;
+import org.apache.log4j.helpers.LogLog;
+import org.apache.log4j.spi.LoggingEvent;
+
+/**
+ * ApexRollingFileAppender extends {@link RollingFileAppender}
+ * This appender doesn't keep renaming files when file reach max size.
+ * The appender starts with <filename>.0 file when file reaches max size,
+ * it creates next file <filename>.1 and so on till <filename>.MaxBackupIndex.
+ * After that it will delete <filename>.0 and start creating log files in that
+ * order again.
+ * To identify current file, it creates <filename> symlink to current file.
+ */
+public class ApexRollingFileAppender extends RollingFileAppender
```

Comment:
I discussed with @vrozov briefly about this, and we think there should be an interface that has two methods. E.g.:
```java
interface LogLocationProvider
{
  String getCurrentFileName();
  long getCurrentFileOffset();
}
```
and have ApexRollingFileAppender implement this interface so that it's possible for developers to implement this interface to have their own file appender that works with this feature. And the code that says `appender instanceof ApexRollingFileAppender` would become `appender instanceof LogLocationProvider`

Note that I'm not using the name getLastExceptionOffset because it's really the current offset of the log file, and has nothing to do with exceptions.